src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -391,6 +391,41 @@ public interface IQuantumProcessor
         /// <param name="qubits">Register to apply the exponent to.</param>
         void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits);
 
+        /// <summary>
+        /// Used as extensibility mechanism to add custom gates to IQuantumProcessor. 
+        /// </summary>
+        /// <remarks>
+        /// The custom gate may use only some of the parameters. Depending on the nature of the gate, all of them may be optional. 
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="gateId">The id of the custom gate. It is up to the custom gate implementor to define the meaning.</param>
+        /// <param name="qubits">Array of qubits to apply the gate to. Their meaning is specific to the custom gate.</param>
+        /// <param name="paulis">Array of single-qubit Pauli values. Their meaning is specific to the custom gate.</param>
+        /// <param name="longs">Array of long parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="doubles">Array of double parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="bools">Array of bool parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="results">Array of Result parameters. Their meaning is specific to the custom gate.</param>
+        /// <param name="strings">Array of string parameters. Their meaning is specific to the custom gate.</param>
+        void Gate(long gateId, IQArray<Qubit> qubits, IQArray<Pauli> paulis, IQArray<long> longs, IQArray<double> doubles, IQArray<bool> bools, IQArray<Result> results, IQArray<string> strings);
```

Comment:
A default implementation is provided in a QuantumProcessorBase class. While formally this is a breaking change, none of known implementations will be broken on this, since all of them derive from the QuantumProcessorBase class. In general, I would strongly recommend against implementations in interfaces. And in this case it can be easily avoided. I bet nobody would notice (be broken) on this, and we are not yet at version 1.0, so a breaking change like this is acceptable,

---
In reply to: [410516089](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r410516089) [](ancestors = 410516089)